udata/models/url_field.py

```diff
@@ -1,14 +1,80 @@
 # -*- coding: utf-8 -*-
 from __future__ import unicode_literals
 
+import re
+
 from mongoengine.fields import URLField as MEURLField
 
 
+IP_MIDDLE_OCTET = r'(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5]))'
+IP_LAST_OCTET = r'(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))'
+
+URL_REGEX = re.compile(
+    r'^'
+    # scheme is validated separately
+    r'^(?:[a-z0-9\.\-]*)://'
+    # user:pass authentication
+    r'(?:\S+(?::\S*)?@)?'
+    r'(?:'
+    r'(?P<private_ip>'
+    # IP address exclusion
+    # private & local networks
+    r'(?:(?:10|127)' + IP_MIDDLE_OCTET + r'{2}' + IP_LAST_OCTET + r')|'
+    r'(?:(?:169\.254|192\.168)' + IP_MIDDLE_OCTET + IP_LAST_OCTET + r')|'
+    r'(?:172\.(?:1[6-9]|2\d|3[0-1])' + IP_MIDDLE_OCTET + IP_LAST_OCTET + r'))'
+    r'|'
+    # IP address dotted notation octets
+    # excludes loopback network 0.0.0.0
+    # excludes reserved space >= 224.0.0.0
+    # excludes network & broadcast addresses
+    # (first & last IP address of each class)
+    r'(?P<public_ip>'
+    r'(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])'
+    r'' + IP_MIDDLE_OCTET + r'{2}'
+    r'' + IP_LAST_OCTET + r')'
+    r'|'
+    # host name
+    r'(?:(?:[a-z\u00a1-\uffff0-9]-?)*[a-z\u00a1-\uffff0-9]+)'
+    # domain name
+    r'(?:\.(?:[a-z\u00a1-\uffff0-9]-?)*[a-z\u00a1-\uffff0-9]+)*'
+    # TLD identifier
+    r'(?:\.(?:[a-z\u00a1-\uffff]{2,}))'
+    r')'
+    # port number
+    r'(?::\d{2,5})?'
+    # resource path
+    r'(?:/\S*)?'
+    # query string
+    r'(?:\?\S*)?'
+    r'$',
+    re.UNICODE | re.IGNORECASE
+)
+
+
 class URLField(MEURLField):
     '''
     An URL field that automatically strips extra spaces
+    and support uncode domain and paths.
+
+    Public URL can be enforced with `public=True`
+
+    URL_REGEX has been extracted and adapted from:
+        https://github.com/kvesteri/validators/blob/master/validators/url.py
```

Comment:
What is different versus the `validators.url` implementation? Public vs private?

Reply:
Schema validation is handled separately (MongoEngine implementation).
Raw strings are used instead of unicode

Reply:
I added it in the comment